plugin/completion.py

```diff
@@ -1,25 +1,85 @@
 from .core.registry import session_for_view, client_from_session, LSPViewEventListener
+from .core.configurations import is_supported_syntax
+from .core.documents import get_document_position
 from .core.sessions import Session
-from .core.settings import settings, client_configs
+from .core.edit import parse_text_edit
+from .core.views import range_to_region
 from .core.typing import Any, List, Dict, Tuple, Optional, Union
 
 
-class CompletionState(object):
-    IDLE = 0
-    REQUESTING = 1
-    APPLYING = 2
-    CANCELLING = 3
+last_text_command = None
 
 
-last_text_command = None
+class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
+    def run(self, edit: 'Any', item: 'Dict') -> None:
+        insert_text_format = item.get("insertTextFormat")
+
+        text_edit = item.get('textEdit')
+        if text_edit:
+            new_text = text_edit.get('newText')
+
+            range = Range.from_lsp(text_edit['range'])
+            edit_region = range_to_region(range, self.view)
+
+            if insert_text_format == InsertTextFormat.Snippet:
+                self.view.run_command("insert_snippet", {"contents": new_text})
+            else:
+                # use current point as the end of the edit region
+                current_point = self.view.sel()[0].begin()
+                edit_range = sublime.Region(edit_region.begin(), current_point)
+                self.view.replace(edit, edit_range, new_text)
+
+                # move the cursor to the end of the text edit
+                sel = self.view.sel()
+                sel.clear()
+                sel.add(edit_range.begin() + len(new_text))
+        else:
+            completion = item.get('insertText') or item.get('label') or ""
+            current_point = self.view.sel()[0].begin()
+            if insert_text_format == InsertTextFormat.Snippet:
+                self.view.run_command("insert_snippet", {"contents": completion})
+            else:
+                self.view.insert(edit, current_point, completion)
+
+        # import statements, etc. some servers only return these after a resolve.
+        additional_edits = item.get('additionalTextEdits')
+        if additional_edits:
+            self.apply_additional_edits(additional_edits)
+
+        self.do_resolve(item)
+
+    def do_resolve(self, item: dict) -> None:
+        session = session_for_view(self.view, 'completionProvider', self.view.sel()[0].begin())
+        if not session:
+            return
+
+        client = client_from_session(session)
+        if not client:
+            return
+
+        completion_provider = session.get_capability('completionProvider')
+        has_resolve_provider = completion_provider and completion_provider.get('resolveProvider', False)
+        if has_resolve_provider:
+            client.send_request(Request.resolveCompletionItem(item), self.handle_resolve_response)
+
+    def handle_resolve_response(self, response: 'Optional[Dict]') -> None:
```

Comment:
You don't have to put quotes around these mypy types anymore. Just write `Optional[Dict]` ! :dancer: